pkg/threescale/connectors/backend.go

```diff
@@ -0,0 +1,141 @@
+package connectors
+
+import (
+	"fmt"
+	"time"
+
+	backend "github.com/3scale/3scale-go-client/client"
+	"github.com/3scale/3scale-istio-adapter/pkg/threescale/metrics"
+)
+
+type AuthRepResult string
+
+const Unknown AuthRepResult = "UNKNOWN"
+const Ok AuthRepResult = "AUTHENTICATED"
+const Unauthenticated AuthRepResult = "UNAUTHENTICATED"
+const cacheMiss AuthRepResult = "UNCACHED"
+
+const (
+	UserKey = iota + 1
+	AppId
+)
+
+type authRepFn func(auth backend.TokenAuth, key string, svcID string, params backend.AuthRepParams, extension map[string]string) (backend.ApiResponse, error)
+
+type AuthRepUserAuth struct {
+	// Api Key Authn Pattern
+	UserKey string
+	// App Id Authn Pattern - Mutually exclusive with Api Key Pattern
+	AppID string
+	// Optional App Key as a key pair for App ID pattern
+	AppKey   string
+	AuthType int
+}
+
+type AuthRepRequest struct {
+	ServiceID string
+	UserAuth  AuthRepUserAuth
+	TokenAuth backend.TokenAuth
+	Metrics   backend.Metrics
+}
+
+type authRepRemoteCall struct {
+	authKey        string
+	params         backend.AuthRepParams
+	callback       authRepFn
+	remoteHostname string
+}
+
+type BackendCache interface {
+	AuthRep(svcID string, metrics backend.Metrics) AuthRepResult
+	Set()
+}
+
+type BackendConnector struct {
+	cache           BackendCache
+	reportMetricsFn metrics.ReportMetricsFn
+}
+
+type LocalCache struct {
+}
+
+func (l LocalCache) AuthRep(svcID string, metrics backend.Metrics) AuthRepResult {
+	return cacheMiss
+}
+
+func (l LocalCache) Set() {
+
+}
+
+func NewBackendConnector() *BackendConnector {
+	return &BackendConnector{}
+}
+
+func (bc *BackendConnector) DoAuthRep(req AuthRepRequest, metrics backend.Metrics, c *backend.ThreeScaleClient) (backend.ApiResponse, error) {
+	var resp backend.ApiResponse
+
+	callRemote := func() (backend.ApiResponse, error) {
+		remoteReq, err := bc.authRepRemoteReqBuilder(req.UserAuth, metrics, c)
+		if err != nil {
+			return resp, err
+		}
+
+		return bc.callRemote(req.TokenAuth, remoteReq, req.ServiceID)
+	}
+
+	if bc.cache != nil {
+		switch result := bc.cache.AuthRep("", nil); result {
+		case cacheMiss:
+			resp, err := callRemote()
+			if err != nil {
+				bc.cache.Set()
+			}
+			return resp, err
+		case Unauthenticated:
+			resp.Success = false
+			resp.Reason = "limits exceeded"
+			resp.StatusCode = 429
+
+		case Ok:
+			resp.StatusCode = 200
+			resp.Success = true
+		}
+		return resp, nil
+
+	} else {
+		return callRemote()
+	}
+
+	return resp, nil
+}
+
+func (bc *BackendConnector) callRemote(tokenAuth backend.TokenAuth, req authRepRemoteCall, svcID string) (backend.ApiResponse, error) {
+	const endpoint = "AuthRep"
+	const target = "Backend"
+	var (
+		start   time.Time
+		elapsed time.Duration
+	)
+
+	start = time.Now()
+	resp, apiErr := req.callback(tokenAuth, req.authKey, svcID, req.params, nil)
```

Comment:
I think we could eliminate here this callback by just receiving the needed information in the parameters and calling 3scale provided we modify the function above (see comment above re: callback).

Reply:
Solved with new structure